File: backend/app/api/routes/geo.py
from typing import Any

from fastapi import APIRouter, HTTPException, Query

from app.api.deps import SessionDep
from app.repositories.geo_repository import GeoDataUnavailableError
from app.schemas.geo import GeoOption, IucnOption
from app.schemas.response import PagedApiResponse, api_page
from app.services import geo_service

router = APIRouter(prefix="/geo", tags=["地理字典 / Geo Dictionary"])

@router.get(
    "/gadm",
    response_model=PagedApiResponse[list[GeoOption]],
    summary="获取 GADM 行政区划选项 / Get GADM administrative options",
)
def get_gadm_options(
    session: SessionDep,
    level: int = Query(..., description="行政级别: 0=国家, 1=省/州, 2=市/县 / Level 0=Country, 1=Province, 2=City"),
    parent_gid: str | None = Query(None, description="上级行政区 GID，用于联动过滤 / Parent GID for cascading filter"),
    search: str | None = Query(None, description="关键字模糊搜索 / Keyword search"),
    page: int = Query(1, ge=1, description="页码 / Page number"),
    page_size: int = Query(100, ge=1, le=100, description="每页条数 / Page size"),
) -> Any:
    """
    获取 GADM 全局行政区划选项。 / Get GADM administrative boundary options.

    支持关键字模糊搜索，支持通过 parent_gid (上级 GID) 进行下级联动查询。
    Supports keyword search and cascading filtering via parent GID.
    无需身份验证。 / No authentication required.
    """
    try:
        data, total = geo_service.get_gadm_options(
            session=session,
            level=level,
            parent_gid=parent_gid,
            search=search,
            page=page,
            page_size=page_size,
        )
    except GeoDataUnavailableError as e:
        raise HTTPException(status_code=503, detail=str(e))
    return api_page(data=data, total=total, page=page, page_size=page_size)


@router.get(
    "/iho",
    response_model=PagedApiResponse[list[GeoOption]],
    summary="获取 IHO 海域选项 / Get IHO Sea Area options",
)
def get_iho_options(
    session: SessionDep,
    search: str | None = Query(None, description="关键字模糊搜索 / Keyword search"),
    page: int = Query(1, ge=1, description="页码 / Page number"),
    page_size: int = Query(100, ge=1, le=100, description="每页条数 / Page size"),
) -> Any:
    """
    获取 IHO 全球海域选项。 / Get IHO globally sea area options.

    无需身份验证。 / No authentication required.
    """
    data, total = geo_service.get_iho_options(
        session=session,
        search=search,
        page=page,
        page_size=page_size,
    )
    return api_page(data=data, total=total, page=page, page_size=page_size)


@router.get(
    "/iucn-realms",
    response_model=PagedApiResponse[list[IucnOption]],
    summary="获取 IUCN 界别选项 / Get IUCN Realm options",
)
def get_iucn_realms(
    session: SessionDep,
    search: str | None = Query(None, description="关键字搜索 / Keyword search"),
    page: int = Query(1, ge=1, description="页码 / Page number"),
    page_size: int = Query(100, ge=1, le=100, description="每页条数 / Page size"),
) -> Any:
    """
    获取 IUCN 全球生态系统分类 Level 1 (Realm) 界别选项。
    Get IUCN Global Ecosystem Typology Level 1 (Realm) options.

    无需身份验证。 / No authentication required.
    """
    data, total = geo_service.get_iucn_realms(
        session=session,
        search=search,
        page=page,
        page_size=page_size,
    )
    return api_page(data=data, total=total, page=page, page_size=page_size)


@router.get(
    "/iucn-biomes",
    response_model=PagedApiResponse[list[IucnOption]],
    summary="获取 IUCN 群系选项 / Get IUCN Biome options",
)
def get_iucn_biomes(
    session: SessionDep,
    realm_id: int | None = Query(None, description="所属界别 ID / Parent Realm ID"),
    search: str | None = Query(None, description="关键字搜索 / Keyword search"),
    page: int = Query(1, ge=1, description="页码 / Page number"),
    page_size: int = Query(100, ge=1, le=100, description="每页条数 / Page size"),
) -> Any:
    """
    获取 IUCN 全球生态系统分类 Level 2 (Biome) 群系选项。
    Get IUCN Global Ecosystem Typology Level 2 (Biome) options.

    支持根据 realm_id (Level 1) 进行联动过滤。
    Supports cascading filtering by realm_id (Level 1).
    无需身份验证。 / No authentication required.
    """
    data, total = geo_service.get_iucn_biomes(
        session=session,
        realm_id=realm_id,
        search=search,
        page=page,
        page_size=page_size,
    )
    return api_page(data=data, total=total, page=page, page_size=page_size)


@router.get(
    "/iucn-functional-types",
    response_model=PagedApiResponse[list[IucnOption]],
    summary="获取 IUCN 功能组选项 / Get IUCN Functional Type options",
)
def get_iucn_functional_types(
    session: SessionDep,
    biome_id: int | None = Query(None, description="所属群系 ID / Parent Biome ID"),
    search: str | None = Query(None, description="关键字搜索 / Keyword search"),
    page: int = Query(1, ge=1, description="页码 / Page number"),
    page_size: int = Query(100, ge=1, le=100, description="每页条数 / Page size"),
) -> Any:
    """
    获取 IUCN 全球生态系统分类 Level 3 (Functional Group) 功能组选项。
    Get IUCN Global Ecosystem Typology Level 3 (Functional Group) options.

    支持根据 biome_id (Level 2) 进行联动过滤。
    Supports cascading filtering by biome_id (Level 2).
    无需身份验证。 / No authentication required.
    """
    data, total = geo_service.get_iucn_functional_types(
        session=session,
        biome_id=biome_id,
        search=search,
        page=page,
        page_size=page_size,
    )
    return api_page(data=data, total=total, page=page, page_size=page_size)

File: frontend/src/api/endpoints/geo.ts
import { apiClient } from "../client"
import type { PagedApiResponse } from "../../types"

/** Matches GET /v1/geo/gadm items: { gid, name } */
export interface GadmOption {
    gid: string
    name: string
    name_zh?: string
}

/** IUCN endpoints use `id`; IHO uses `gid` (string of DB id). */
export interface GenericGeoOption {
    name: string
    id?: number
    gid?: string
}

export type GeoPageParams = {
    search?: string
    page?: number
    page_size?: number
}

export type CoordinateGeoOption = { gid: string; name: string }
export type CoordinateMatches = {
    gadm: {
        status: "matched" | "unmatched" | "ambiguous"
        gadm0: CoordinateGeoOption | null
        gadm1: CoordinateGeoOption | null
        gadm2: CoordinateGeoOption | null
    }
    iho: {
        status: "matched" | "unmatched" | "ambiguous"
        option: CoordinateGeoOption | null
    }
}

export const geoApi = {
    getCoordinateMatches(longitude: number, latitude: number, ignoreUnauthorized?: boolean) {
        return apiClient.get<{ code: number; message?: string; data?: CoordinateMatches }>("/v1/geo/coordinate-matches", {
            params: { longitude, latitude },
            ignoreUnauthorized,
        })
    },
    /** 
     * 获取 GADM 行政区划选项 / Get GADM administrative options
     * @param params - Query parameters
     * @param params.level - integer (Level) 0=Country, 1=Province, 2=City
     * @param params.parent_id - Parent Id (integer) or undefined
     * @param params.search - Search (string) or undefined Keyword search
     */
    getGadmOptions(
        params: GeoPageParams & { level: number; parent_gid?: string },
        ignoreUnauthorized?: boolean,
    ) {
        return apiClient.get<PagedApiResponse<GadmOption[]>>("/v1/geo/gadm", { params, ignoreUnauthorized })
    },

    getIhoOptions(params?: GeoPageParams, ignoreUnauthorized?: boolean) {
        return apiClient.get<PagedApiResponse<GenericGeoOption[]>>("/v1/geo/iho", { params, ignoreUnauthorized })
    },

    getIucnRealms(params?: GeoPageParams, ignoreUnauthorized?: boolean) {
        return apiClient.get<PagedApiResponse<GenericGeoOption[]>>("/v1/geo/iucn-realms", {
            params,
            ignoreUnauthorized,
        })
    },

    getIucnBiomes(params?: GeoPageParams & { realm_id?: number }, ignoreUnauthorized?: boolean) {
        const clean = params
            ? (Object.fromEntries(
                  Object.entries(params).filter(([, v]) => v !== undefined && v !== ""),
                  ) as GeoPageParams & { realm_id?: number })
            : undefined
        return apiClient.get<PagedApiResponse<GenericGeoOption[]>>("/v1/geo/iucn-biomes", {
            params: clean,
            ignoreUnauthorized,
        })
    },

    getIucnFunctionalTypes(params?: GeoPageParams & { biome_id?: number }, ignoreUnauthorized?: boolean) {
        const clean = params
            ? (Object.fromEntries(
                  Object.entries(params).filter(([, v]) => v !== undefined && v !== ""),
                  ) as GeoPageParams & { biome_id?: number })
            : undefined
        return apiClient.get<PagedApiResponse<GenericGeoOption[]>>("/v1/geo/iucn-functional-types", {
            params: clean,
            ignoreUnauthorized,
        })
    },
}
